CMakeLists.txt

```diff
@@ -0,0 +1,217 @@
+
+add_library(overlay SHARED
+  overlay/messages.pb.cc
+  overlay/overlay.pb.cc
+  overlay/agent_metrics.cpp
+  overlay/agent.cpp
+  overlay/supervisor_metrics.cpp
+  overlay/master_metrics.cpp
+  overlay/master.cpp
+)
+install(TARGETS overlay LIBRARY DESTINATION lib/mesos)
+
+add_custom_command(
+  OUTPUT overlay/messages.pb.cc overlay/overlay.pb.cc
+  COMMAND protoc
+    -I${CMAKE_CURRENT_SOURCE_DIR}
+    --cpp_out=${CMAKE_CURRENT_BINARY_DIR}
+    overlay/messages.proto overlay/overlay.proto
+  DEPENDS overlay/messages.proto overlay/overlay.proto
+)
+
+target_include_directories(overlay PRIVATE
+  "${CMAKE_CURRENT_SOURCE_DIR}"
+  "${CMAKE_CURRENT_BINARY_DIR}"
+  "${BOOST_INCL}"
+  "${GLOG_INCL}"
+  "${PROTOBUF_INCL}"
+  "${PICOJSON_INCL}"
+  "${RAPIDJSON_INCL}"
+  "${PROCESS_INCL}"
+  "${STOUT_INCL}"
+  "${CURL_INCL}"
+  "${ZOOKEEPER_INCL}"
+  "${MESOS_ROOT}/build/include"
+  "${MESOS_ROOT}/include"
+)
+
+
+# metrics
+
+add_library(metrics SHARED
+  metrics/messages.pb.cc
+  metrics/isolator.cpp
+)
+install(TARGETS metrics
+  ARCHIVE DESTINATION lib/mesos
+  LIBRARY DESTINATION lib/mesos
+)
+
+add_custom_command(
+  OUTPUT metrics/messages.pb.cc
+  COMMAND protoc
+    -I${CMAKE_CURRENT_SOURCE_DIR}
+    --cpp_out=${CMAKE_CURRENT_BINARY_DIR}
+    metrics/messages.proto
+  DEPENDS metrics/messages.proto
+)
+
+target_include_directories(metrics PRIVATE
+  "${CMAKE_CURRENT_SOURCE_DIR}"
+  "${CMAKE_CURRENT_BINARY_DIR}"
+  "${BOOST_INCL}"
+  "${CURL_INCL}"
+  "${GLOG_INCL}"
+  "${MESOS_ROOT}/build/include"
+  "${MESOS_ROOT}/include"
+  "${PICOJSON_INCL}"
+  "${PROCESS_INCL}"
+  "${PROTOBUF_INCL}"
+  "${RAPIDJSON_INCL}"
+  "${STOUT_INCL}"
+)
+
+if (BUILD_TESTING)
+  add_subdirectory(tests)
+
+  # Test binary for metrics
+  add_executable(test-metrics
+    metrics/messages.pb.cc
+    tests/metrics_tests.cpp
+  )
+
+  # TODO(akornatskyy): Paths defined as "nowhere" are needed to build Mesos
+  # test utils (https://github.com/apache/mesos/blob/master/src/tests/utils.cpp).
+  # These definitions can be dropped if the test utilities we need here are
+  # split from the path-dependent utilities into a separate file.
+  add_definitions(
+    -DMODULES_BUILD_DIR="${PROJECT_BINARY_DIR}"
+    -DSOURCE_DIR="${MESOS_ROOT}"
+    -DBUILD_DIR="${MESOS_ROOT}/build"
+    -DPKGMODULEDIR="nowhere"
+    -DLIBDIR="nowhere"
+    -DPKGLIBEXECDIR="nowhere"
+    -DTESTLIBEXECDIR="nowhere"
+    -DSBINDIR="nowhere"
+    -DPKGDATADIR="nowhere"
+    -DVERSION="1.10.0"
+  )
+
+  target_include_directories(test-metrics PRIVATE
+    "${CMAKE_CURRENT_SOURCE_DIR}"
+    "${CMAKE_CURRENT_BINARY_DIR}"
+    "${BOOST_INCL}"
+    "${CURL_INCL}"
+    "${GLOG_INCL}"
+    "${GMOCK_INCL}"
+    "${GTEST_INCL}"
+    "${MESOS_ROOT}/build/include"
+    "${MESOS_ROOT}/build/src"
+    "${MESOS_ROOT}/include"
+    "${MESOS_ROOT}/src"
+    "${NVML_INCL}"
+    "${PICOJSON_INCL}"
+    "${PROCESS_INCL}"
+    "${PROTOBUF_INCL}"
+    "${RAPIDJSON_INCL}"
+    "${STOUT_INCL}"
+    "${ZOOKEEPER_INCL}"
```

Comment:
Can you compose the list of include dirs for tests once and reuse it for all the test binaries?
like
```
list(APPEND TEST_INCLUDE_DIRS ${MESOS_ROOT}/include ${MESOS_ROOT}/src ...)
...
target_include_directories(test-metrics PRIVATE ${TEST_INCLUDE_DIRS})
```

The same applies to lists of libraries in `target_link_libraries()`

Reply:
yep. I will use the approach in `dcos-ee-mesos-modules`, with a list of includes.